File: 5/p2.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func parseFile(fileName string) []int {
	data, _ := ioutil.ReadFile(fileName)
	program := []int{}
	for _, s := range strings.Split(string(data), ",") {
		trimmed := strings.Trim(s, "\n")
		if i, err := strconv.Atoi(trimmed); err == nil {
			program = append(program, i)
		}
	}
	return program
}

type Memory struct {
	data []int
}

type Mode int

const Immediate = 1
const Position = 0

type OpCode int

const Add = 1
const Multiply = 2
const Input = 3
const Output = 4
const JumpIfTrue = 5
const JumpIfFalse = 6
const LessThan = 7
const Equals = 8
const Halt = 99

func (m *Memory) read(position int, mode Mode) int {
	switch mode {
	case Immediate:
		return m.data[position]
	case Position:
		return m.data[m.data[position]]
	}
	return 0
}

func (m *Memory) write(position int, value int) {
	m.data[m.data[position]] = value
}

func parseInstruction(instruction int) (OpCode, Mode, Mode) {
	opCode := OpCode(instruction % 100)
	modeA := Mode((instruction % 1000) / 100)
	modeB := Mode((instruction % 10000) / 1000)
	return opCode, modeA, modeB
}

func runProgram(memory Memory, input int) int {
	position := 0
	output := 0

	for {
		instruction := memory.read(position, Immediate)
		opCode, modeA, modeB := parseInstruction(instruction)

		switch opCode {
		case Halt:
			return output
		case Add:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			memory.write(position+3, a+b)
			position += 4
		case Multiply:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			memory.write(position+3, a*b)
			position += 4
		case Input:
			memory.write(position+1, input)
			position += 2
		case Output:
			output = memory.read(position+1, modeA)
			position += 2
		case JumpIfTrue:
			a := memory.read(position+1, modeA)
			if a != 0 {
				position = memory.read(position+2, modeB)
			} else {
				position += 3
			}
		case JumpIfFalse:
			a := memory.read(position+1, modeA)
			if a == 0 {
				position = memory.read(position+2, modeB)
			} else {
				position += 3
			}
		case LessThan:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			if a < b {
				memory.write(position+3, 1)
			} else {
				memory.write(position+3, 0)
			}
			position += 4
		case Equals:
			a := memory.read(position+1, modeA)
			b := memory.read(position+2, modeB)
			if a == b {
				memory.write(position+3, 1)
			} else {
				memory.write(position+3, 0)
			}
			position += 4
		}
	}
}

func main() {
	program := parseFile("program.testcode")
	m := Memory{data: program}
	result := runProgram(m, 1)
	fmt.Println(result)
}

File: 5/p2.py
def parse_file(filename):
    with open(filename) as f:
        code = f.readline()
        ints = code.split(',')
        program = [int(i) for i in ints]
        return program

def digit(instruction, place):
    code_map = '{:05d}'.format(instruction)
    return int(code_map[place])

def read(memory, position, mode):
    if mode == 1:
        return memory[position]
    if mode == 0:
        return memory[memory[position]]

def write(memory, position, value):
    memory[memory[position]] = value

def run_program(memory, program_input):
    position = 0
    output = None

    while True:
        code = memory[position]
        mode_a = digit(code, 2)
        mode_b = digit(code, 1)
        op_code = code % 100

        if op_code == 99:
            return output

        if op_code == 1:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            write(memory, position + 3, a + b)
            position += 4

        if op_code == 2:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            write(memory, position + 3, a * b)
            position += 4

        if op_code == 3:
            write(memory, position + 1, program_input)
            position += 2

        if op_code == 4:
            output = read(memory, position + 1, mode_a)
            position += 2

        if op_code == 5:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            if a != 0:
                position = b
            else:
                position += 3

        if op_code == 6:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            if a == 0:
                position = b
            else:
                position += 3

        if op_code == 7:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            if a < b:
                write(memory, position + 3, 1)
            else:
                write(memory, position + 3, 0)
            position += 4

        if op_code == 8:
            a = read(memory, position + 1, mode_a)
            b = read(memory, position + 2, mode_b)
            if a == b:
                write(memory, position + 3, 1)
            else:
                write(memory, position + 3, 0)
            position += 4


    return output

program = parse_file('program.testcode')
print run_program(program, 5)
